tensorboard/webapp/app_routing/route_contexted_reducer_helper.ts

```diff
@@ -110,62 +110,96 @@ export function createRouteContextedState<
     privateRouteContextedState: {},
   };
 
+  /**
+   * Updates routeful state to match key `afterNamespaceId`. Takes existing routeful
+   * state and caches it using key `beforeNamespaceId`.
+   */
+  function updateRoutefulState(
+    state: FullState,
+    beforeRouteId: string | null,
+    beforeNamespaceId: string | null,
+    // TODO(bdubois): afterNamespaceId should not allow null when it is no
+    // longer optional.
+    afterNamespaceId: string | null
+  ) {
+    let nextContextedStateCache: {
+      [routeId: string]: RoutefulState;
+    } = {...state.privateRouteContextedState};
+
+    if (beforeNamespaceId) {
+      // Swap out routeful state to cache, keyed by beforeNamespaceId.
+      const routefulStateToCache = {} as RoutefulState;
+      for (const key of keys) {
+        routefulStateToCache[key] = (state as RoutefulState)[key];
+      }
+      nextContextedStateCache = {
+        ...nextContextedStateCache,
+        [beforeNamespaceId]: routefulStateToCache,
+      };
+    }
+
+    // Update routeful state to reflect afterNamespaceId.
+    let nextRoutefulState = {};
+    // Note: state.privateRouteContextedState always exists in practice except
+    // for in tests.
+    if (
+      afterNamespaceId &&
+      state.privateRouteContextedState?.[afterNamespaceId]
+    ) {
+      // Swap in existing state since it already exists in the cache.
+      nextRoutefulState = state.privateRouteContextedState[afterNamespaceId];
+    } else if (beforeRouteId) {
+      // Reset to initial state since we had a cache miss and this is not
+      // initial load.
+      //
+      // Note: We don't reset to initial state on initial load because we
+      // assume the state already has values from bootstrapping deeplinks and
+      // we should not overwrite the values.
+      nextRoutefulState = routefulInitialState;
+    }
+
+    return {
+      ...state,
+      ...nextRoutefulState,
+      privateRouteContextedState: nextContextedStateCache,
```

Comment:
Would you mind renaming `privateRouteContextedState` to `privateNamespacedState` in this PR? Since it is a private property, it should be safe to do it here.